ProcFilter.h

```diff
@@ -16,72 +16,106 @@
 #ifndef AUOMS_PROC_FILTER_H
 #define AUOMS_PROC_FILTER_H
 
-#include<sys/time.h>
+#include <sys/time.h>
 #include <string>
 #include <memory>
 #include <set>
 #include <unordered_set>
 #include <list>
 #include <unordered_map>
 #include <queue>
+#include <regex>
 #include "Config.h"
 #include "UserDB.h"
 #include "ProcessInfo.h"
+#include "ProcessDefines.h"
 
-struct ProcInfo {
-    int pid;
-    int ppid;
-    int uid;
-    std::string exe;
-    std::string arg1;
+enum StringMatchType { MatchUndefined, MatchEquals, MatchStartsWith, MatchContains, MatchRegex };
 
-    ProcInfo(ProcessInfo* proc);
-    static const ProcInfo Empty;
-
-    inline bool operator==(const ProcInfo& x) const;
-    inline bool operator!=(const ProcInfo& x) const;
+struct cmdlineFilter {
+    enum StringMatchType _matchType;
+    std::string _matchValue;
+    std::regex _matchRegex;
 };
 
+bool operator==(const cmdlineFilter& a, const cmdlineFilter& b);
+
 struct ProcFilterSpec {
-    ProcFilterSpec(const std::string& exe, const std::string& arg1, uint32_t flags, const std::string& user) {
-        _exe = exe;
-        _arg1 = arg1;
-        _flags = flags;
-        _user = user;
+    ProcFilterSpec(uint32_t match_mask, int depth, int uid, int gid,
+            std::vector<std::string>& syscalls, const std::string& exeMatchValue,
+            const std::vector<cmdlineFilter>& cmdlineFilters)
+    {
+        _match_mask = match_mask;
+        _depth = depth;
+        _uid =  uid;
+        _gid = gid;
+        std::copy(syscalls.cbegin(), syscalls.cend(), std::inserter(_syscalls, _syscalls.end()));
+        _exeMatchValue = exeMatchValue;
+        if (match_mask & PFS_MATCH_EXE_REGEX) {
+            _exeRegex = std::regex(exeMatchValue, std::regex::optimize);
+        }
+        for (auto cf : cmdlineFilters) {
+            if (cf._matchType == MatchRegex) {
+                cf._matchRegex = std::regex(cf._matchValue, std::regex::optimize);
+            }
+            _cmdlineFilters.emplace_back(cf);
+        }
+    }
+
+    uint32_t _match_mask;
+    int _depth;
+    uint32_t _uid;
+    uint32_t _gid;
+    std::vector<std::string> _syscalls;
```

Comment:
Please consider using a bitset instead. Iterating over a vector for each syscall event is much more CPU intensive than doing one syscall to bitset index lookup, then doing a bitset position test. The syscall name to bitset position map can be stored in ProceFilter as a mutex protected static member.

You would also need to add a seperate boolean indicating if the bitset is include or exclude.

Reply:
I appreciate that there is a performance consideration with this data structure. I propose that I will revisit this in a new PR as a straight swap for a bitset isn't sufficient - currently the order matters due to the way the syscalls are presented and while it is perfectly possible to replace with a bitset, I'd like to spend some time developing and testing in order to get it right. In the mean time I'd like this PR to progress and be merged. 
Your concerns are noted and appreciated.

Reply:
I revisted this and replaced the vector with an unordered_map which should provide quick look-ups.